File: sudoku.js
// sudoku.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const readline = require('readline');

const COLORS = {
    reset: '\x1b[0m',
    red: '\x1b[91m',
    green: '\x1b[92m',
    yellow: '\x1b[93m',
    blue: '\x1b[94m',
    cyan: '\x1b[96m',
    gray: '\x1b[90m',
    bold: '\x1b[1m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

class Sudoku {
    constructor(board = null) {
        this.board = board ? board.map(row => [...row]) : Array.from({ length: 9 }, () => Array(9).fill(0));
        this.steps = 0;
    }

    render(highlight = null) {
        const lines = [];
        for (let i = 0; i < 9; i++) {
            if (i % 3 === 0 && i > 0) {
                lines.push(colorize('┃━━━┃━━━┃━━━┃', 'gray'));
            }
            const row = [];
            for (let j = 0; j < 9; j++) {
                if (j % 3 === 0 && j > 0) {
                    row.push(colorize('┃', 'gray'));
                }
                const val = this.board[i][j];
                if (val === 0) {
                    row.push(' ');
                } else {
                    let col = 'green';
                    if (highlight && highlight.some(h => h[0] === i && h[1] === j)) {
                        col = 'cyan';
                    }
                    row.push(colorize(String(val), col));
                }
            }
            lines.push(row.join(' '));
        }
        return lines.join('\n');
    }

    parseString(str) {
        if (str.length !== 81) throw new Error('Строка должна содержать ровно 81 символ');
        for (let i = 0; i < 9; i++) {
            for (let j = 0; j < 9; j++) {
                const ch = str[i*9 + j];
                if (ch >= '1' && ch <= '9') {
                    this.board[i][j] = parseInt(ch, 10);
                } else if (ch === '0') {
                    this.board[i][j] = 0;
                } else {
                    throw new Error(`Недопустимый символ: ${ch}`);
                }
            }
        }
    }

    exportString() {
        let s = '';
        for (let i = 0; i < 9; i++) {
            for (let j = 0; j < 9; j++) {
                s += this.board[i][j] === 0 ? '0' : String(this.board[i][j]);
            }
        }
        return s;
    }

    isValid() {
        // Строки
        for (let i = 0; i < 9; i++) {
            const seen = new Set();
            for (let j = 0; j < 9; j++) {
                const val = this.board[i][j];
                if (val !== 0) {
                    if (val < 1 || val > 9 || seen.has(val)) return false;
                    seen.add(val);
                }
            }
        }
        // Столбцы
        for (let j = 0; j < 9; j++) {
            const seen = new Set();
            for (let i = 0; i < 9; i++) {
                const val = this.board[i][j];
                if (val !== 0) {
                    if (seen.has(val)) return false;
                    seen.add(val);
                }
            }
        }
        // Блоки
        for (let br = 0; br < 3; br++) {
            for (let bc = 0; bc < 3; bc++) {
                const seen = new Set();
                for (let i = 0; i < 3; i++) {
                    for (let j = 0; j < 3; j++) {
                        const val = this.board[br*3+i][bc*3+j];
                        if (val !== 0) {
                            if (seen.has(val)) return false;
                            seen.add(val);
                        }
                    }
                }
            }
        }
        return true;
    }

    getCandidates(row, col) {
        if (this.board[row][col] !== 0) return [];
        const used = new Set();
        for (let j = 0; j < 9; j++) if (this.board[row][j] !== 0) used.add(this.board[row][j]);
        for (let i = 0; i < 9; i++) if (this.board[i][col] !== 0) used.add(this.board[i][col]);
        const br = Math.floor(row/3)*3, bc = Math.floor(col/3)*3;
        for (let i = 0; i < 3; i++) {
            for (let j = 0; j < 3; j++) {
                const val = this.board[br+i][bc+j];
                if (val !== 0) used.add(val);
            }
        }
        const cands = [];
        for (let v = 1; v <= 9; v++) if (!used.has(v)) cands.push(v);
        return cands;
    }

    findBestEmpty() {
        let bestR = -1, bestC = -1, bestCands = null, minCands = 10;
        for (let i = 0; i < 9; i++) {
            for (let j = 0; j < 9; j++) {
                if (this.board[i][j] === 0) {
                    const cands = this.getCandidates(i, j);
                    if (cands.length < minCands) {
                        minCands = cands.length;
                        bestR = i; bestC = j; bestCands = cands;
                        if (minCands === 1) return [bestR, bestC, bestCands];
                    }
                }
            }
        }
        return [bestR, bestC, bestCands];
    }

    solve(animate = false, delay = 0.1) {
        this.steps = 0;
        const start = Date.now();
        const _solve = () => {
            this.steps++;
            const [r, c, cands] = this.findBestEmpty();
            if (r === -1) return true;
            for (const val of cands) {
                this.board[r][c] = val;
                if (animate) this.printAnimated(r, c, delay);
                if (_solve()) return true;
                this.board[r][c] = 0;
            }
            return false;
        };
        const solved = _solve();
        const elapsed = (Date.now() - start) / 1000;
        return [solved, elapsed];
    }

    printAnimated(row, col, delay) {
        console.clear();
        console.log(this.render([[row, col]]));
        console.log(colorize(`Шаг: ${this.steps}`, 'yellow'));
        // Задержка
        const stop = Date.now() + delay * 1000;
        while (Date.now() < stop) {}
    }

    countSolutions(maxCount = 100) {
        let count = 0;
        const _count = () => {
            if (count >= maxCount) return;
            const [r, c, cands] = this.findBestEmpty();
            if (r === -1) { count++; return; }
            for (const val of cands) {
                this.board[r][c] = val;
                _count();
                this.board[r][c] = 0;
                if (count >= maxCount) return;
            }
        };
        _count();
        return count;
    }

    static generate(level = 'easy') {
        const levels = { easy: [30, 35], medium: [40, 45], hard: [50, 55] };
        const [minEmpty, maxEmpty] = levels[level] || levels.easy;
        // Заполняем доску
        const s = new Sudoku();
        s.solveInternal();
        const filled = s.board.map(row => [...row]);
        const target = minEmpty + Math.floor(Math.random() * (maxEmpty - minEmpty + 1));
        let removed = 0, attempts = 0;
        while (removed < target && attempts < 10000) {
            attempts++;
            const i = Math.floor(Math.random() * 9);
            const j = Math.floor(Math.random() * 9);
            if (filled[i][j] !== 0) {
                const backup = filled[i][j];
                filled[i][j] = 0;
                const test = new Sudoku(filled);
                if (test.countSolutions(2) === 1) {
                    removed++;
                } else {
                    filled[i][j] = backup;
                }
            }
        }
        return new Sudoku(filled);
    }

    solveInternal() {
        const _solve = () => {
            const [r, c, cands] = this.findBestEmpty();
            if (r === -1) return true;
            for (const val of cands) {
                this.board[r][c] = val;
                if (_solve()) return true;
                this.board[r][c] = 0;
            }
            return false;
        };
        _solve();
    }
}

function loadBoard(source) {
    let content;
    if (fs.existsSync(source)) {
        content = fs.readFileSync(source, 'utf8');
    } else {
        content = source;
    }
    const digits = content.split('').filter(ch => ch >= '0' && ch <= '9');
    if (digits.length !== 81) throw new Error('Должно быть ровно 81 цифра');
    const s = new Sudoku();
    s.parseString(digits.join(''));
    return s;
}

function main() {
    const args = process.argv.slice(2);
    if (args.length < 1) {
        console.log(colorize('Usage: node sudoku.js <generate|solve|check|count|export> [options]', 'yellow'));
        console.log('  generate -l <easy|medium|hard>');
        console.log('  solve -i <file|string> [-a] [--delay <sec>]');
        console.log('  check -i <file|string>');
        console.log('  count -i <file|string> [--max <N>]');
        console.log('  export -i <file|string>');
        process.exit(1);
    }

    const cmd = args[0];
    const opts = {};
    for (let i = 1; i < args.length; i++) {
        if (args[i] === '-l' && i+1 < args.length) opts.level = args[++i];
        else if (args[i] === '-i' && i+1 < args.length) opts.input = args[++i];
        else if (args[i] === '-o' && i+1 < args.length) opts.output = args[++i];
        else if (args[i] === '-a') opts.animate = true;
        else if (args[i] === '--delay' && i+1 < args.length) opts.delay = parseFloat(args[++i]);
        else if (args[i] === '--max' && i+1 < args.length) opts.max = parseInt(args[++i]);
    }

    try {
        switch (cmd) {
            case 'generate': {
                const level = opts.level || 'easy';
                const s = Sudoku.generate(level);
                if (opts.output) fs.writeFileSync(opts.output, s.exportString());
                console.log(s.render());
                const empty = s.board.flat().filter(v => v === 0).length;
                console.log(colorize(`Уровень: ${level}, пустых ячеек: ${empty}`, 'yellow'));
                break;
            }
            case 'solve': {
                if (!opts.input) throw new Error('Укажите входную доску через -i');
                const s = loadBoard(opts.input);
                if (!s.isValid()) throw new Error('Доска невалидна!');
                const [solved, elapsed] = s.solve(opts.animate, opts.delay || 0.1);
                if (solved) {
                    if (opts.output) fs.writeFileSync(opts.output, s.exportString());
                    console.log(s.render());
                    console.log(colorize(`Решение найдено за ${elapsed.toFixed(3)} сек, шагов: ${s.steps}`, 'green'));
                } else {
                    console.log(colorize('Решения не существует!', 'red'));
                }
                break;
            }
            case 'check': {
                if (!opts.input) throw new Error('Укажите входную доску через -i');
                const s = loadBoard(opts.input);
                console.log(s.isValid() ? colorize('✅ Доска корректна.', 'green') : colorize('❌ Доска невалидна.', 'red'));
                break;
            }
            case 'count': {
                if (!opts.input) throw new Error('Укажите входную доску через -i');
                const s = loadBoard(opts.input);
                const maxCount = opts.max || 100;
                const count = s.countSolutions(maxCount);
                if (count >= maxCount) {
                    console.log(colorize(`Количество решений >= ${maxCount} (ограничено)`, 'yellow'));
                } else {
                    console.log(colorize(`Количество решений: ${count}`, 'green'));
                }
                break;
            }
            case 'export': {
                if (!opts.input) throw new Error('Укажите входную доску через -i');
                const s = loadBoard(opts.input);
                console.log(s.exportString());
                break;
            }
            default:
                console.log(colorize(`Неизвестная команда: ${cmd}`, 'red'));
        }
    } catch (err) {
        console.log(colorize(`Ошибка: ${err.message}`, 'red'));
        process.exit(1);
    }
}

main();

File: sudoku.rb
#!/usr/bin/env ruby
# sudoku.rb
# encoding: UTF-8

require 'json'
require 'fileutils'
require 'timeout'

COLORS = {
  reset: "\e[0m",
  red: "\e[91m",
  green: "\e[92m",
  yellow: "\e[93m",
  blue: "\e[94m",
  cyan: "\e[96m",
  gray: "\e[90m",
  bold: "\e[1m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

class Sudoku
  attr_reader :board, :steps

  def initialize(board = nil)
    @board = board ? board.map(&:dup) : Array.new(9) { Array.new(9, 0) }
    @steps = 0
  end

  def render(highlight = nil)
    lines = []
    9.times do |i|
      lines << colorize('┃━━━┃━━━┃━━━┃', :gray) if i % 3 == 0 && i > 0
      row = []
      9.times do |j|
        row << colorize('┃', :gray) if j % 3 == 0 && j > 0
        val = @board[i][j]
        if val == 0
          row << ' '
        else
          col = :green
          if highlight && highlight.include?([i, j])
            col = :cyan
          end
          row << colorize(val.to_s, col)
        end
      end
      lines << row.join(' ')
    end
    lines.join("\n")
  end

  def parse_string(str)
    raise "Строка должна содержать 81 символ" if str.size != 81
    9.times do |i|
      9.times do |j|
        ch = str[i*9 + j]
        if ch.between?('1', '9')
          @board[i][j] = ch.to_i
        elsif ch == '0'
          @board[i][j] = 0
        else
          raise "Недопустимый символ: #{ch}"
        end
      end
    end
  end

  def export_string
    @board.flatten.map { |v| v == 0 ? '0' : v.to_s }.join
  end

  def valid?
    # Строки
    9.times do |i|
      seen = {}
      9.times do |j|
        val = @board[i][j]
        next if val == 0
        return false if val < 1 || val > 9 || seen[val]
        seen[val] = true
      end
    end
    # Столбцы
    9.times do |j|
      seen = {}
      9.times do |i|
        val = @board[i][j]
        next if val == 0
        return false if seen[val]
        seen[val] = true
      end
    end
    # Блоки
    3.times do |br|
      3.times do |bc|
        seen = {}
        3.times do |i|
          3.times do |j|
            val = @board[br*3 + i][bc*3 + j]
            next if val == 0
            return false if seen[val]
            seen[val] = true
          end
        end
      end
    end
    true
  end

  def candidates(row, col)
    return [] if @board[row][col] != 0
    used = {}
    9.times { |j| used[@board[row][j]] = true if @board[row][j] != 0 }
    9.times { |i| used[@board[i][col]] = true if @board[i][col] != 0 }
    br, bc = row / 3 * 3, col / 3 * 3
    3.times do |i|
      3.times do |j|
        val = @board[br+i][bc+j]
        used[val] = true if val != 0
      end
    end
    (1..9).reject { |v| used[v] }
  end

  def find_best_empty
    best_r = best_c = -1
    best_cands = nil
    min_cands = 10
    9.times do |i|
      9.times do |j|
        if @board[i][j] == 0
          cands = candidates(i, j)
          if cands.size < min_cands
            min_cands = cands.size
            best_r, best_c = i, j
            best_cands = cands
            return [best_r, best_c, best_cands] if min_cands == 1
          end
        end
      end
    end
    [best_r, best_c, best_cands]
  end

  def solve_internal
    r, c, cands = find_best_empty
    return true if r == -1
    cands.each do |val|
      @board[r][c] = val
      return true if solve_internal
      @board[r][c] = 0
    end
    false
  end

  def solve(animate = false, delay = 0.1)
    @steps = 0
    start = Time.now
    solve_fn = -> {
      @steps += 1
      r, c, cands = find_best_empty
      return true if r == -1
      cands.each do |val|
        @board[r][c] = val
        if animate
          system('clear') || system('cls')
          puts render([[r, c]])
          puts colorize("Шаг: #{@steps}", :yellow)
          sleep(delay)
        end
        return true if solve_fn.call
        @board[r][c] = 0
      end
      false
    }
    solved = solve_fn.call
    elapsed = Time.now - start
    [solved, elapsed]
  end

  def count_solutions(max_count = 100)
    count = 0
    count_fn = -> {
      return if count >= max_count
      r, c, cands = find_best_empty
      if r == -1
        count += 1
        return
      end
      cands.each do |val|
        @board[r][c] = val
        count_fn.call
        @board[r][c] = 0
        return if count >= max_count
      end
    }
    count_fn.call
    count
  end

  def self.generate(level = 'easy')
    levels = { 'easy' => [30, 35], 'medium' => [40, 45], 'hard' => [50, 55] }
    min_empty, max_empty = levels[level] || [30, 35]

    s = Sudoku.new
    s.solve_internal
    filled = s.board.map(&:dup)
    target = rand(min_empty..max_empty)
    removed = 0
    attempts = 0
    while removed < target && attempts < 10000
      attempts += 1
      i, j = rand(9), rand(9)
      if filled[i][j] != 0
        backup = filled[i][j]
        filled[i][j] = 0
        test = Sudoku.new(filled)
        if test.count_solutions(2) == 1
          removed += 1
        else
          filled[i][j] = backup
        end
      end
    end
    Sudoku.new(filled)
  end
end

def load_board(source)
  content = File.exist?(source) ? File.read(source) : source
  digits = content.chars.select { |ch| ch.between?('0', '9') }
  raise "Должно быть 81 цифра" if digits.size != 81
  s = Sudoku.new
  s.parse_string(digits.join)
  s
end

def main
  if ARGV.empty?
    puts colorize("Usage: ruby sudoku.rb <generate|solve|check|count|export> [options]", :yellow)
    puts "  generate -l <easy|medium|hard>"
    puts "  solve -i <file|string> [-a] [--delay <sec>]"
    puts "  check -i <file|string>"
    puts "  count -i <file|string> [--max <N>]"
    puts "  export -i <file|string>"
    exit 1
  end

  cmd = ARGV[0]
  input = nil
  output = nil
  level = 'easy'
  animate = false
  delay = 0.1
  max_count = 100

  i = 1
  while i < ARGV.size
    case ARGV[i]
    when '-i' then input = ARGV[i+1]; i += 1
    when '-o' then output = ARGV[i+1]; i += 1
    when '-l' then level = ARGV[i+1]; i += 1
    when '-a' then animate = true
    when '--delay' then delay = ARGV[i+1].to_f; i += 1
    when '--max' then max_count = ARGV[i+1].to_i; i += 1
    end
    i += 1
  end

  begin
    case cmd
    when 'generate'
      s = Sudoku.generate(level)
      File.write(output, s.export_string) if output
      puts s.render
      empty = s.board.flatten.count(0)
      puts colorize("Уровень: #{level}, пустых ячеек: #{empty}", :yellow)

    when 'solve'
      raise "Укажите входную доску через -i" unless input
      s = load_board(input)
      raise "Доска невалидна!" unless s.valid?
      solved, elapsed = s.solve(animate, delay)
      if solved
        File.write(output, s.export_string) if output
        puts s.render
        puts colorize("Решение найдено за #{elapsed.round(3)} сек, шагов: #{s.steps}", :green)
      else
        puts colorize("Решения не существует!", :red)
      end

    when 'check'
      raise "Укажите входную доску через -i" unless input
      s = load_board(input)
      puts s.valid? ? colorize("✅ Доска корректна.", :green) : colorize("❌ Доска невалидна.", :red)

    when 'count'
      raise "Укажите входную доску через -i" unless input
      s = load_board(input)
      cnt = s.count_solutions(max_count)
      if cnt >= max_count
        puts colorize("Количество решений >= #{max_count} (ограничено)", :yellow)
      else
        puts colorize("Количество решений: #{cnt}", :green)
      end

    when 'export'
      raise "Укажите входную доску через -i" unless input
      s = load_board(input)
      puts s.export_string

    else
      puts colorize("Неизвестная команда: #{cmd}", :red)
    end
  rescue => e
    puts colorize("Ошибка: #{e.message}", :red)
    exit 1
  end
end

main if __FILE__ == $0
